databuilder/query_model.py

```diff
@@ -365,38 +359,83 @@ class DomainMismatchError(Exception):
     ...
 
 
-PATIENT_DOMAIN = object()
+def validate_input_domains(node):
+    # The domain of a Frame or Series can be thought of as the set of its primary keys.
+    # This determines which other Frames or Series it can be validly composed with. The
+    # rules we want to enforce are:
+    #
+    #  * One-row-per-patient data can be composed with anything: by design every table
+    #    has a patient_id column and so data defined purely in terms of patient_id
+    #    can be joined with anything.
+    #
+    #  * Many-rows-per-patient data drawn from different tables can never be combined
+    #    — at least, not until we add an explicit JOIN operation.
+    #
+    #  * Many-rows-per-patient data drawn from a single underlying table can be combined
+    #    as long as any filters applied to the data can be arranged in a linear
+    #    sequence. That is, it's legal to combine `Table A -> Filter B -> Filter C` and
+    #    `Table A -> Filter B`. But it's not legal to combine `Table A -> Filter D` and
+    #    `Table A -> Filter E`. There's never a need to use such constructions and they
+    #    will almost always be the result of user error. It's better to reject them
+    #    immediately than give the user unexpected results at query time.
```

Comment:
Since this comment is going to be the main way people new to the code have to get their heads round this, I think that an example showing why an illegal construct would be wrong would be helpful.